tests/integration/setup.js

```diff
@@ -0,0 +1,520 @@
+}
+
+export function givenLoggedInUser(beforeFn) {
+	beforeFn(() =>
+		balena.auth
+			.login({
+				email: exports.credentials.email,
+				password: exports.credentials.password,
+			})
+			.then(exports.resetUser),
+	);
+
+	const afterFn = beforeFn === beforeEach ? afterEach : after;
+	return afterFn(() => exports.resetUser());
+}
+
+export function loginPaidUser() {
+	return balena.auth.login({
+		email: exports.credentials.paid.email,
+		password: exports.credentials.paid.password,
+	});
+}
+
+const resetApplications = () =>
+	balena.pine.delete({
+		resource: 'application',
+		options: {
+			$filter: { 1: 1 },
+		},
+	});
+
+export function givenInitialOrganization(beforeFn) {
+	return beforeFn(function () {
+		return getInitialOrganization().then((initialOrg) => {
+			return (this.initialOrg = initialOrg);
+		});
+	});
+}
+
+export function givenAnApplication(beforeFn) {
+	exports.givenInitialOrganization(beforeFn);
+
+	beforeFn(function () {
+		// calling this.skip() doesn't trigger afterEach,
+		// so we need to reset in here as well
+		// See: https://github.com/mochajs/mocha/issues/3740
+		return resetApplications()
+			.then(() => {
+				return balena.models.application.create({
+					name: 'FooBar',
+					applicationType: 'microservices-starter',
+					deviceType: 'raspberry-pi',
+					organization: this.initialOrg.id,
+				});
+			})
+			.then((application) => {
+				this.application = application;
+				chai
+					.expect(this.application.is_for__device_type)
+					.to.be.an('object')
+					.that.has.property('__id')
+					.that.is.a('number');
+
+				return balena.pine.get({
+					resource: 'device_type',
+					id: this.application.is_for__device_type.__id,
+					options: {
+						$select: 'slug',
+					},
+				});
+			})
+			.then((applicationDeviceType) => {
+				this.applicationDeviceType = applicationDeviceType;
+				return chai
+					.expect(this.applicationDeviceType)
+					.to.be.an('object')
+					.that.has.property('slug')
+					.that.is.a('string');
+			});
+	});
+
+	const afterFn = beforeFn === beforeEach ? afterEach : after;
+	return afterFn(resetApplications);
+}
+
+const resetDevices = () =>
+	balena.pine.delete({
+		resource: 'device',
+		options: {
+			$filter: { 1: 1 },
+		},
+	});
+
+export function givenADevice(beforeFn, extraDeviceProps) {
+	beforeFn(function () {
+		// calling this.skip() doesn't trigger afterEach,
+		// so we need to reset in here as well
+		// See: https://github.com/mochajs/mocha/issues/3740
+		return resetDevices()
+			.then(() => {
+				const uuid = balena.models.device.generateUniqueKey();
+				return balena.models.device.register(this.application.app_name, uuid);
+			})
+			.tap((deviceInfo) => {
+				if (!this.currentRelease || !this.currentRelease.commit) {
+					return;
+				}
+
+				return balena.pine.patch({
+					resource: 'device',
+					body: {
+						is_running__release: this.currentRelease.id,
+					},
+					options: {
+						$filter: {
+							uuid: deviceInfo.uuid,
```

Comment:
@thgreasi this fails to type check because of the magic around body props which means you can only filter on props that are also included in the body - how are we meant to get around those typings with force casts (which are very hard to do in javascript+prettier) and ideally without just `@ts-ignore`ing it

Reply:
:disappointed: I should have set `AnyObject` as the default typing...
The only reason I didn't do so, is that consumers will have to explicitly provide the type, in order to have this type checked, while if I indeed had `= AnyObject`, we could have untyped pine calls all over the place, just b/c we forgot to provide the type.
In the next major the SDK will have a list of known resources and pick the type auto-magically :)